[WIP] Add subnet validator unit tests

matte21 opened the pull request:
#### What's currently here
A subset of the subnet validator unit test cases, intended to serve as an example to drive the discussion on what the unit tests will look like.

#### Rationale for chosen testing style
Unlike many upstream controllers' unit tests, these tests don't express expected output in terms of API calls the controller makes, rather, desired output is expressed in terms of existing subnets at the end of the test (this is similar to what the Persistent Volume controller does). This has the advantage of being more declarative and high-level, and exposing less of the implementation. Another difference is that these tests `Run()` the controller, the rationale is again hiding implementation details and having tests that are more declarative. Follows an explanation. Assume that a more conventional testing style was adopted, where API objects are manually fed for processing to the controller under test. Assume that you wanted to test that a subnet S1 with a rival S2 is not validated when processed. You would create a fake client that mocks an API server, and populate it with S1 and S2, you would then create an informer, pass the fake client to it, and add S1 and S2 to the informer's store. You would then invoke something like `validator.processSubnet(S1)` and check whether the last API call performed with the fake client is an update to S1 which marks it as valid; the informer would not be `Start()`ed and the controller would not be `Run()`. But in the current validator implementation that would not work, because processing of S1 would fail. So for the test to be meaningful S2 would have to be processed first. This already is ugly and couples the tests too much (IMO) with the implementation.

#### Drawbacks of chosen testing style

The fact that controller and informer are `Run()` and `Start()`ed respectively, means that there's no certain and automatic way to determine when a test has ended. The tests periodically retrieve the whole subnet population with the fake client and compare it to the expected one. A mismatch could be due to a bug or to the fact that the validator is correct but has not yet produced the expected output. To avoid tests that hang indefinitely, after a fixed number of comparisons between got and expected subnets the test fails, but this could be due to the validator being slow => that's a spurious failure. Also, there's a lower bound to each test completion because of how test outcome is evaluated (e.g. wait a little bit and then compare expected and existing subnets); if tests are run sequentially they can be slow. Also, because the informer is  `Start()`ed and has multiple goroutines, different test runs with the same inputs lead to different executions of the validator, making failures less reproducible.

matte21 committed:
Add first draft of subnet validator unit tests

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -229,15 +230,27 @@ func NewValidationController(netIfc kosclientv1a1.NetworkV1alpha1Interface,
+		// Register Prometheus metrics only if it's not a test because during a
+		// test the subnet validator constructor might be invoked multiple times
+		// and each invocation attempts to register the same Prometheus metrics
+		// names, which panics.
```
An alternate way to solve the problem would be to create and register these in an init() function.

Regarding the eventRecorder: the solution here works, but the usual pattern is to have a more general constructor that takes an eventRecorder as a parameter.

matte21 replied:
Having a test flag in the constructor is ugly, so both suggestions were followed in 7ce6161 .

Concerning the `eventRecorder` as a constructor arg, I'm not sure whether you meant adding another constructor (used by the less general one) --- 7ce6161 doesn't do that, it simply modifies the existing constructor to accept the `eventRecorder` as a parameter (and removes the args whose sole purpose is setting up the `eventRecorder`). The real `eventRecorder` is created in the kos controller manager function which starts the validator (https://github.com/MikeSpreitzer/kube-examples/pull/83/files#diff-bbc277f7a59a5d8649a40e4007252ea2R102-R105); this is different wrt what some core K8s controllers do, i.e., creating the `eventRecorder` directly in the constructor and setting a fake one during tests simply by setting the appropriate field of the controller's struct. I chose the current approach because it was less code wrt to having more constructors, because it doesn't needlessly instantiate a real `eventRecorder` during tests, and because it leads to fewer constructor arguments.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+	// function that claims equality between `Status.Errors` of two subnets in a
+	// more liberal way.
+	cmpSubnetStatusErrs := cmp.Comparer(func(errs1, errs2 []string) bool {
+		return (len(errs1) > 0 && len(errs2) > 0) || (len(errs1) == 0 && len(errs2) == 0)
```
Would `set.NewString(errs1...).Equal(set.NewString(errs2...))` be an upgrade?  (`import "k8s.io/apimachinery/pkg/util/sets"`.)

matte21 replied:
Suggestion followed in 7ce6161 . Notice that this also requires explicitly stating the status error messages of the expected subnets for every test case, 7ce6161 also adds the functional options to do that (https://github.com/MikeSpreitzer/kube-examples/pull/83/files#diff-dae925a38a8446d84d44a6492a14613eR487-R499).

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+
+		// go-cmp considers slices equals only if they contain the same elements
+		// in the same order, but we have no control on the order of the slice
+		// of subnets which represents the test output. We don't want spurius
```
s/spurius/spurious/

matte21 replied:
Fixed in 7ce6161 .

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+		subnetsInformer.Informer(),
+		subnetsInformer.Lister(),
+		nil,
+		// Use a fake rate limiter (delay is always 0) to reduce likelyhood of
```
"...likelihood of spurious..."

matte21 replied:
Fixed in 7ce6161 .

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+		Duration: 500 * time.Millisecond,
+		Factor:   0.8,
+		Steps:    10,
+	}
```
See the sort of testing done in https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go , where the active goroutines are all accounted for so the test funcion knows when all the work is done.  If we could extend that accounting through the workqueue then we could do it here too and not have to worry about when to sample the state and when it is settled.

MikeSpreitzer replied:
The problem with workqueue is that the number of Pop does not necessarily equal the number of Add + AddRateLimited.  But in our test cases it probably will, so we _could_ accurately track the number of active goroutines.

MikeSpreitzer replied:
Actually, AddRateLimited has the added complication that it unblocks a goroutine in the future.  Doing that accurately would likely involve EventClock::EventAfterDuration.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+				},
+			},
+		},
+
```
How about a test case for the same VNI being used in two namespaces?

matte21 replied:
Yes, I plan to add more tests than those currently here, which only act as an initial core to drive discussion on the general outline of unit tests for the subnet validator (and to a lesser extent for the other kos controllers).

matte21 replied:
I just noticed though that in [this](https://github.com/MikeSpreitzer/kube-examples/pull/81#discussion_r337855636) comment we discussed removing the constraint that a VNI must be confined to a single K8s API namespace. If indeed that's what we plan to do, we might skip adding unit tests that make sure that constraint is enforced.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+	subnet := update.GetObject().(*netv1a1.Subnet)
+	rv, _ := strconv.Atoi(subnet.ResourceVersion)
+	rv++
+	subnet.ResourceVersion = strconv.Itoa(rv)
```
Parsing a ResourceVersion is gauche.  How about simply appending a digit?

matte21 replied:
Fixed in 7ce6161 .

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator_test.go:
```diff
@@ -0,0 +1,506 @@
+}
+
+//****************************************************************************
+// Follow functional options and util functions to create test subnets with  //
```
"The following functional options and util functions create ..."

matte21 replied:
Fixed in 7ce6161 .

matte21 committed:
Implement some suggestions from the code review

1) Fix typos in comments.

2) Change the way resource version is changed on updates in a way
which does not require parsing it (which is awkward).

3) Move set-up of Prometheus metrics to init function.

4) Make K8s event recorder a constructor arg so that unit tests
can pass in a fake recorder. This (together with point 3) allows
getting rid of the ugly testing flag in the subnet validator
controller's constructor.

5) Test status error messages more precisely.

6) (not in the review comments) Fix a bug in a unit test: the subnet
validator was set up to have 0 worker goroutines instead of 1.

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -333,25 +333,27 @@ func (v *Validator) OnDelete(obj interface{}) {
+		return false
+	}
+	subnetNSN := queueItem.(k8stypes.NamespacedName)
+	defer v.queue.Done(subnetNSN)
```
I would put the defer as soon as possible (i.e., one line earlier).

MikeSpreitzer commented on staging/kos/pkg/controllers/subnet/validator.go:
```diff
@@ -60,57 +58,8 @@ const (
-	staleRVsMutex sync.RWMutex
-
+// Prometheus metrics.
+var (
```
Some day we may want to test with fake Prometheus instrumentation (https://github.com/prometheus/client_golang/blob/master/prometheus/testutil/testutil.go#L17-L24).
But today is not that day.
